packages/fx-core/src/plugins/resource/cicd/plugin.ts

```diff
@@ -0,0 +1,84 @@
+import { Inputs, v2, Result, FxError, Platform } from "@microsoft/teamsfx-api";
+import { FxCICDPluginResultFactory as ResultFactory } from "./result";
+import { CICDProviderFactory } from "./providers/factory";
+import { ProviderKind } from "./providers/enums";
+import { questionNames } from "./questions";
+import { InternalError, NoProjectOpenedError } from "./errors";
+import { generateBuildScript } from "./utils/buildScripts";
+import { telemetryHelper } from "./utils/telemetry-helper";
+
+export class CICDImpl {
+  public async addCICDWorkflows(
+    context: Context,
+    inputs: Inputs,
+    envInfo: v2.EnvInfoV2
+  ): Promise<Result<any, FxError>> {
+    // 1. Key inputs (envName, provider, template) x (hostingType, ).
+    if (!inputs.projectPath) {
+      throw new NoProjectOpenedError();
+    }
+    const projectPath = inputs.projectPath;
+    // By default(VSC), get env name from plugin's own `target-env` question.
+    let envName = inputs[questionNames.Environment];
+    if (inputs.platform == Platform.CLI) {
+      // In CLI, get env name from the default `env` question.
+      envName = envInfo.envName;
+    }
+    const providerName = inputs[questionNames.Provider];
+    const templateNames = inputs[questionNames.Template] as string[];
+    if (!envName || !providerName || templateNames.length == 0) {
+      throw new InternalError("Some preconditions of inputs are not met.");
+    }
+
+    telemetryHelper.sendSuccessEvent(context, envInfo, "collect-cicd-questions", {
+      env: envName,
+      provider: providerName,
+      template: templateNames.join(","),
+    });
+
+    // 2. Call factory to get provider instance.
+    const providerInstance = CICDProviderFactory.create(providerName as ProviderKind);
+
+    // 3. Call instance.scaffold(template, replacements: any).
+    //  3.1 Construct replacements.
+    const hostType = context.projectSetting.solutionSettings?.hostType;
+    const replacements = {
+      env_name: envName,
+      build_script: generateBuildScript(context.projectSetting),
+      hosting_type_contains_spfx: hostType == "SPFx",
+      hosting_type_contains_azure: hostType == "Azure",
+      cloud_resources_contains_sql:
+        context.projectSetting.solutionSettings?.["azureResources"].includes("sql") ?? false,
+      api_prefix: context.projectSetting.appName,
+      cloud_resources_contains_apim:
+        context.projectSetting.solutionSettings?.["azureResources"].includes("apim") ?? false,
+    };
+
+    const progressBar = context.userInteraction.createProgressBar(
+      "Scaffolding workflow automation files",
+      templateNames.length
+    );
+
+    await progressBar.start(`Scaffolding workflow file for ${templateNames[0]}.`);
+    await providerInstance.scaffold(projectPath, templateNames[0], replacements);
+
+    //  3.2 Call scaffold.
+    for (const templateName of templateNames.slice(1)) {
+      await progressBar.next(`Scaffolding workflow file for ${templateName}.`);
```

Comment:
How long does it take to scaffold from a template? Do we really need to make each of them a independent step?
Or could we make them parallel?

Reply:
It's pretty quick and it's unnecessary to make it parallel.
Another point to make it sequential is it's good for the progress bar.